### PR TITLE
feat(preset-manager): CRUD operations UI

### LINKED ISSUE #24: Stepper buttons remain focused
### How can we get in touch with you if we need more info?

_No response_

### What happened?

Stepper buttons remain focused and selected after click

### Version

Alpha

### Waze Environment

Production (www.waze.com/editor)

### What browsers are you seeing the problem on?

Chrome

### Additional Context

_No response_

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by davidsl4**
This pull request delivers foundational elements for the Preset CRUD (Create, Read, Update, Delete) User Interface. It incorporates the initial setup for preset creation, introduces significant enhancements to our stepper components for improved type-safety and state management, and includes a key bug fix.

## Development Status and Roadmap

The changes included in this pull request mark the successful completion of the initial phase for the Preset CRUD feature, focusing on establishing the necessary foundation and the preliminary "Create Preset" workflow.

### Key deliverables in this pull request include
* The establishment of the project structure for the Preset CRUD UI
* The implementation of the "Create Preset" button, which triggers a `ModalStepper`
* The definition of steps (with icons) and the data structure within this `ModalStepper` (currently with visual placeholders for forms)
* The configuration of `ClosurePresetsListContext` to allow mutable operations by setting `isReadOnly` to `false`
* A significant enhancement to the `ModalStepper` with a generic Data type for improved type safety
* The introduction of the `createUseStepState` utility for robust, step-specific state management
* And a resolution for the context button focus retention issue (Fixes #24).

### Task Checklist

**Overall Goal:** Implement a full CRUD (Create, Read, Update, Delete) interface for presets.

**Phase 1: Foundation & Create Functionality**
* [x] Reserve project structure for Preset CRUD UI
* [x] Implement "Create Preset" button triggering `ModalStepper`
* [x] Define steps and data structure for preset creation in `ModalStepper` (visual placeholders)
* [x] Set `isReadOnly` to `false` in `ClosurePresetsListContext` while we can't query it from an actual storing strategy
* [x] Enhance `ModalStepper` with generic `Data` type for type safety
* [x] Create `createUseStepState` utility for step-specific state management
* [x] Fix context button focus retention issue (#24)
* [x] Implement actual forms and input fields within each step of the `ModalStepper` for creating presets.
* [x] Connect aggregated preset data from the "Create" stepper to the context method for saving new presets.

**Phase 2: Edit & Delete Functionality**
* [x] Integrate the "Edit preset" action with the Preset CRUD UI defined in Phase 1
* [x] Connect aggregated preset data from the "Edit" stepper to the context method for updating presets.
* [x] Design and implement UI for "Delete Preset" actions (including confirmation).
* [x] Connect "Delete" actions to the context method for removing presets.

### CODE REVIEW

**Comment by davidsl4** [comment]
@coderabbitai review

**Comment by davidsl4** [comment]
@gemini-code-assist review

File: src/components/WeekdayPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      </ToggleGroup>
+    </div>
+  );
+}
+
+function isSingleSelectMode(
+  props: WeekdayPickerProps,
+): props is SingleWeekdayPickerProps {
+  return !props.allowMultiple;
+}
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/WeekdayPicker.tsx:164]: @gemini-code-assist we don't test React components yet

File: src/components/stepper/components/StepperActions.tsx [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
 
-  if (typeof currentStepConfig.actions === 'function')
-    return currentStepConfig.actions();
+  if (typeof currentStepConfig.actions === 'function') {
+    return currentStepConfig.actions(
+      getStepData(currentStepConfig.id) as Parameters<
+        typeof currentStepConfig.actions
+      >[0],
+    );
+  }
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/stepper/components/StepperActions.tsx:16]: Unfortunately that's the same issue as we discussed here --> https://github.com/TheEditorX/wme-closures-plus/pull/25#discussion_r2105307564

[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/stepper/components/StepperActions.tsx:16]: Unfortunately that's the same issue as we discussed here --> https://github.com/TheEditorX/wme-closures-plus/pull/25#discussion_r2105307564

File: src/utils/create-blurrable-handler.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { SyntheticEvent } from 'react';
+
+export function createBlurrableHandler<T extends HTMLElement, E extends Event>(
+  handler: (e: SyntheticEvent<T, E>) => void,
+): (e: SyntheticEvent<T, E>) => void {
+  return (e) => {
+    // Blur the element to prevent focus retaining
+    const target = e.currentTarget;
+    target.blur();
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/utils/create-blurrable-handler.ts:9]: Indeed, but I also have a suggestion: Let's improve it and use optional chaining - that way we will keep it a single line but will not lose any functionality or safety.

File: src/components/stepper/utils/create-use-step-state.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      // eslint-disable-next-line react-hooks/rules-of-hooks
+      const isInitialized = useRef(false);
+
+      if (isInitialized.current) return;
+      isInitialized.current = true;
+
+      if (requestedData || !initialState) return;
+
+      setRequestedData(initialState);
+    })();
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/stepper/utils/create-use-step-state.ts:51]: I know it violates the rule, but I don't really want to introduce a `ref` hook into the root of the component, nor do I want to create a dedicated hook for that. I believe as long as the IIFE follows the rules of hooks itself, we're fine to use hooks inside.

[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/stepper/utils/create-use-step-state.ts:51]: I believe the order remains the same on every render, and that the IIFE doesn't make any difference to it, and is similar if the IIFE were defined as a function itself. Am I wrong?

File: src/utils/create-bitwise-enum-flags-class.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+          return this._value === 0;
+        },
+        // Setter for 'None = true' could mean reset(), 'None = false' is trickier (means set something else?)
+        // For simplicity, 'None' might be read-only or have specific setter logic.
+        // Let's make it read-only for 'None = 0'.
+        enumerable: true,
+        configurable: true,
+      });
+    }
+
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/utils/create-bitwise-enum-flags-class.ts:189]: Well, in that case, you're right. I missed that we're talking about the `if` block. The unit tests, however, still pass through, meaning we don't cover this edge case (that set is not defined, or the get would potentially return `true` for everything if the `has` method doesn't have any safeguards). Can you write respective unit tests for this edge case?

**Comment by davidsl4** [comment]
@coderabbitai review

File: src/utils/is-instance.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+export function isInstance(value: unknown): boolean;
+export function isInstance(
+  value: unknown,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  constructor: new (...args: any[]) => unknown,
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/utils/is-instance.ts:5]: Should we also replace the parameter name in the typescript definition? I'm thinking of updating it only in the implementation but keeping the full name in the typescript types, so IntelliSense will actually mention "constructor" rather than "ctor", which I feel less appropriate in the context.

File: src/components/closure-presets/PresetListCardMenu.tsx [hunk 7]
```diff
@@ -16,10 +18,23 @@ export function PresetListCardMenu({ preset }: PresetListCardMenuProps) {
...
+
+            const handler = async () => {
+              await deletePreset(preset.id);
+            };
+
+            el.addEventListener('click', handler);
+            return () => {
+              el.removeEventListener('click', handler);
+            };
+          }}
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/closure-presets/PresetListCardMenu.tsx:36]: Actually, in React 19, they have finally introduced [cleanup functions for `ref`s](https://react.dev/blog/2024/12/05/react-19#cleanup-functions-for-refs) 🥳, which means it should be completely safe to subscribe to events that way. I would use the `onClick` prop, but unfortunately, it isn't passed down as an event handler by React (presumably), and it requires us to re-imagine the wheel and subscribe by other means.

[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/closure-presets/PresetListCardMenu.tsx:36]: Great! So I'll resolve this comment. By the way, this limitation seems to expand beyond `wz-menu-item` in an unpredictable pattern. I already encountered other `wz-*` elements that don't correctly register handlers using default props (e.g. `onClick`), while others seem to do (e.g. `wz-button`). So it is really important when working with these elements to know if they will actually receive the props or not. We're expected to resolve this issue once Editor X will release the UI kit library that will handle these event handler registrations gracefully, and will provide first degree type safety.

File: src/utils/create-bitwise-enum-flags-class.ts [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+      set: function (this: DynamicBitwiseEnumFlags, value: boolean) {
+        if (value) {
+          this.set(flagNumericValue as NumVal);
+        } else {
+          this.clear(flagNumericValue as NumVal);
+        }
+      },
+      enumerable: true, // So these properties show up in Object.keys(), for...in
+      configurable: true, // Allows them to be deleted or reconfigured if needed
+    });
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/utils/create-bitwise-enum-flags-class.ts:206]: That's true. We're looking for ways to improve it. At the moment, we've marked the property itself with a readonly modifier. Should we also remove this chunk of code for the sake of YAGNI?

File: src/components/closure-presets/PresetListCardMenu.tsx [hunk 9]
```diff
@@ -16,10 +18,23 @@ export function PresetListCardMenu({ preset }: PresetListCardMenuProps) {
...
+        >
           <i className="w-icon w-icon-trash" /> Delete
         </wz-menu-item>
```
[the remark this answers is not in the document]
Comment (davidsl4) [reply src/components/closure-presets/PresetListCardMenu.tsx:39]: We already covered the `ref` cleanup topic in [detail](https://github.com/TheEditorX/wme-closures-plus/pull/25#discussion_r2106297179). Regarding the error handling, we aren't expected to encounter any errors. While the current goal is to store them locally, its true that separation of concerns will kick in here and tell us to implement error handling, but at the moment it just not beneficial. YAGNI right?

**Comment by davidsl4** [comment]
@coderabbitai review
